hazelcast/src/test/java/com/hazelcast/cluster/ClusterUpgradeTest.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cluster;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.internal.cluster.ClusterService;
+import com.hazelcast.internal.cluster.impl.VersionMismatchException;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.version.Version;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+
+import static com.hazelcast.test.TestClusterUpgradeUtils.assertClusterVersion;
+import static com.hazelcast.test.TestClusterUpgradeUtils.upgradeClusterMembers;
+
+/**
+ * Create a cluster, then change cluster version. This test uses artificial version numbers, to avoid relying on current version.
+ */
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class ClusterUpgradeTest extends HazelcastTestSupport {
+
+    static final Version VERSION_2_0_5 = Version.of(2, 0, 5);
+    static final Version VERSION_2_1_0 = Version.of(2, 1, 0);
+    static final Version VERSION_2_1_1 = Version.of(2, 1, 1);
+    static final Version VERSION_2_2_0 = Version.of(2, 2, 0);
+    static final Version VERSION_3_0_0 = Version.of(3, 0, 0);
+
+    static final int CLUSTER_MEMBERS_COUNT = 3;
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    private TestHazelcastInstanceFactory factory = new TestHazelcastInstanceFactory(10);
```

Comment:
minor note:
When `private TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(10)` is used,  `factory` will be tear down automatically by `HazelcastTestSupport.shutdownNodeFactory()`.